rpc/nodedialer: avoid tripping breaker on context errors

ajwerner opened the pull request:
In #34026 we added logic to ensure that the context was not canclled before
calling in to GRPCDial. This change extends that to also avoid calling
breaker.Fail if the returned error was context.Canceled.

The motivation for this is an observation that breakers can trip due to context
cancellation which race with calls to Dial (from say DistSQL). This behavior
was observed after a node died due to an unrelated corruption bug. It appears
that this node failure triggered a context cancellation which then tripped a
breaker which then lead to a different flow to fail which then lead to another
cancellation which seems to have then tripped another breaker. The evidence for
this exact serioes of events is somewhat scant but we do know for certain that
we saw breakers tripped due to context cancelled which seems wrong.

```
ip-172-31-44-174> I190305 14:53:47.387863 150672 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:322  [n6] circuitbreaker: rpc [::]:26257->2 tripped: failed to grpc dial n2 at ip-172-31-34-81:26257: context canceled
```

This change also cosmetically refactors DialInternalClient and Dial to share
some copy-pasted code which was becoming burdensome.

Release note: None

Commit:
rpc/nodedialer: avoid tripping breaker on context errors

In #34026 we added logic to ensure that the context was not canclled before
calling in to GRPCDial. This change extends that to also avoid calling
breaker.Fail if the returned error was context.Canceled.

The motivation for this is an observation that breakers can trip due to context
cancellation which race with calls to Dial (from say DistSQL). This behavior
was observed after a node died due to an unrelated corruption bug. It appears
that this node failure triggered a context cancellation which then tripped a
breaker which then lead to a different flow to fail which then lead to another
cancellation which seems to have then tripped another breaker. The evidence for
this exact serioes of events is somewhat scant but we do know for certain that
we saw breakers tripped due to context cancelled which seems wrong.

```
ip-172-31-44-174> I190305 14:53:47.387863 150672 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:322  [n6] circuitbreaker: rpc [::]:26257->2 tripped: failed to grpc dial n2 at ip-172-31-34-81:26257: context canceled
```

This change also cosmetically refactors DialInternalClient and Dial to share
some copy-pasted code which was becoming burdensome.

Lastly this change adds a bunch of unit testing to the nodedialer package.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35433)
<!-- Reviewable:end -->

tbg reviewed (commented):
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35433)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/rpc/nodedialer/nodedialer.go, line 153 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-L_Hl0yjDpXx3r07FlYO:-L_Hl0yjDpXx3r07FlYP:b-z0bbc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/76c5ea1c44eea8f48ca2cbbc9ed0cb64d39f525b/pkg/rpc/nodedialer/nodedialer.go#L153)):*
> ```Go
> 		// that a context cancellation might imply something is amiss with the
> 		// connection it is not a clear enough signal. We will however let
> 		// ErrDeadlineExceeded trip the breaker.
> ```

Why? `context.DeadlineExceeded` just means that the incoming context has a deadline or timeout. Why would that inform on the node's status? We don't want this code to trip the breaker:

```go
ctx, cancel := context.WithTimeout(ctx, time.Second)
time.Sleep(990*time.Millisecond)
dialer.Dial(ctx, ...)
```

Maybe the right thing to do here is to only trip the breaker on gRPC errors (`status.FromError(x)`, probably in conjunction with https://github.com/cockroachdb/cockroach/blob/447fe7d6023a3db7473445944ae3693f0f286d17/pkg/storage/queue.go#L772-L777 to unwind causers should they be used somewhere between grpc and this callsite).

Then the code here would read

```go
if isGRPCError(err) && ctx.Err() == nil {
...
}
```

We should make sure that  our gRPC calls internally uses a sane connection timeout (I don't think the incoming ctx generally has one, so that's not a new want).

Taking a step back, I'm less confident that we need to grep out the grpc-originating errors because what other errors could there be, and wouldn't we want to circuit break on them as well?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Thanks for taking a look. I'm going to write some tests now. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-:-L_Ih5-H4HgsEGPTrK80:bklpl3d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/rpc/nodedialer/nodedialer.go, line 153 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-L_Hl0yjDpXx3r07FlYO:-L_IcW4C4QyZKChBHH5B:b-li1hhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/76c5ea1c44eea8f48ca2cbbc9ed0cb64d39f525b/pkg/rpc/nodedialer/nodedialer.go#L153)):*
> Taking a step back, I'm less confident that we need to grep out the grpc-originating errors because what other errors could there be, and wouldn't we want to circuit break on them as well?

I agree with this idea. I've updated the diff to just record into the breaker if `ctx.Err() == nil`

There are a few other errors that could arise but roughly all of them relate to the node shutting down. We could get context errors from the underlying rpc.Context's masterCtx which gets cancelled upon shutdown. Furthermore we could just generally get errors from the stopper itself. None of those are really very interesting though because the node is already shutting down. 


<!-- Sent from Reviewable.io -->

tbg commented:
Sounds good, ping me when this is RFAL

On Wed, Mar 6, 2019 at 4:52 PM ajwerner <notifications@github.com> wrote:

> *@ajwerner* commented on this pull request.
>
> Thanks for taking a look. I'm going to write some tests now.
>
> *Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/35433#-:-L_Ih5-H4HgsEGPTrK80:bklpl3d>*
> status: [image: :shipit:] complete! 0 of 0 LGTMs obtained (waiting on @tbg
> <https://github.com/tbg>)
>
> ------------------------------
>
> *pkg/rpc/nodedialer/nodedialer.go, line 153 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/35433#-L_Hl0yjDpXx3r07FlYO:-L_IcW4C4QyZKChBHH5B:b-li1hhl>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/76c5ea1c44eea8f48ca2cbbc9ed0cb64d39f525b/pkg/rpc/nodedialer/nodedialer.go#L153>):*
>
> Taking a step back, I'm less confident that we need to grep out the
> grpc-originating errors because what other errors could there be, and
> wouldn't we want to circuit break on them as well?
>
> I agree with this idea. I've updated the diff to just record into the
> breaker if ctx.Err() == nil
>
> There are a few other errors that could arise but roughly all of them
> relate to the node shutting down. We could get context errors from the
> underlying rpc.Context's masterCtx which gets cancelled upon shutdown.
> Furthermore we could just generally get errors from the stopper itself.
> None of those are really very interesting though because the node is
> already shutting down.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/35433#pullrequestreview-211311174>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135Nrx9-KKmdjQszMYei-OhxDTOpMjks5vT-RVgaJpZM4bflfq>
> .
>

ajwerner reviewed (commented):
Testing added, RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-:-L_K-B8_7cr3awMwBJft:b-cks6te)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

ajwerner commented:
One bummer about this testing is that it takes 3 seconds due to the hard-coded heartbeat interval in the rpc context. I've typed out code to plumb a setting through which enables making the test much shorter. Is that worth it?

tbg reviewed (approved):
> I've typed out code to plumb a setting through which enables making the test much shorter. Is that worth it?

Yeah, I think so. (For example, we'll get many more iterations during nightly stress that way).

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-:-L_NyHa_0knVD5cXtVUX:b-htopbr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @tbg)

___
*[pkg/rpc/nodedialer/nodedialer.go, line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-L_NwkDXBfTOGWCzuM-P:-L_NwkDXBfTOGWCzuM-Q:bv3pack) ([raw file](https://github.com/cockroachdb/cockroach/blob/72aeb7f3b58cfc74126a69ecbf10121621f0fb56/pkg/rpc/nodedialer/nodedialer.go#L142)):*
> ```Go
> 	defer func() {
> 		// Enforce a minimum interval between warnings for failed connections.
> 		if err != nil && err != ctx.Err() && breaker.ShouldLog() {
> ```

is this the right condition? If the ctx times out while trying to connect, we'll get some gRPC error back (wrapping a context error) and this will trivially be true. I'd go with `if err != nil && ctx.Err() == nil && breaker.ShouldLog() { ... }`
___
*[pkg/rpc/nodedialer/nodedialer_test.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-L_NxXeJ0RM8ZATf1qbn:-L_NxXeJ0RM8ZATf1qbo:bjh5716) ([raw file](https://github.com/cockroachdb/cockroach/blob/72aeb7f3b58cfc74126a69ecbf10121621f0fb56/pkg/rpc/nodedialer/nodedialer_test.go#L53)):*
> ```Go
> 	_, err := nd.Dial(ctx, 1)
> 	assert.Nil(t, err, "failed to dial")
> 	assert.True(t, breaker.Ready())
> ```

Would the breaker trip right away after one attempt? Want to also assert breaker.Failures == 0?
___
*[pkg/rpc/nodedialer/nodedialer_test.go, line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-L_Ny91Y0ivziLMZctaa:-L_Ny91Y0ivziLMZctab:b-px7xdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/72aeb7f3b58cfc74126a69ecbf10121621f0fb56/pkg/rpc/nodedialer/nodedialer_test.go#L172)):*
> becomes


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35433)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35433)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/rpc/nodedialer/nodedialer.go, line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-L_NwkDXBfTOGWCzuM-P:-L_Nyf9RCFAt09tSa4qJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/72aeb7f3b58cfc74126a69ecbf10121621f0fb56/pkg/rpc/nodedialer/nodedialer.go#L142)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

is this the right condition? If the ctx times out while trying to connect, we'll get some gRPC error back (wrapping a context error) and this will trivially be true. I'd go with `if err != nil && ctx.Err() == nil && breaker.ShouldLog() { ... }`
</blockquote></details>

Done.
___
*[pkg/rpc/nodedialer/nodedialer_test.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-L_NxXeJ0RM8ZATf1qbn:-L_Nz8_h2wXz50mFKVxJ:bn1fxr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/72aeb7f3b58cfc74126a69ecbf10121621f0fb56/pkg/rpc/nodedialer/nodedialer_test.go#L53)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Would the breaker trip right away after one attempt? Want to also assert breaker.Failures == 0?
</blockquote></details>

Done. Our threshold is 1 failure for tripping. 

Honestly this test was more to make sure that the testing infrastructure I set up wasn't busted, it doesn't really exercise much beyond that. 
___
*[pkg/rpc/nodedialer/nodedialer_test.go, line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-L_Ny91Y0ivziLMZctaa:-L_O-7Sw8Dvg5ObwJ179:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/72aeb7f3b58cfc74126a69ecbf10121621f0fb56/pkg/rpc/nodedialer/nodedialer_test.go#L172)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> becomes
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
I still need to plumb the config. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-:-L_O2NBPCx6PqY7qhP9T:bz9rbno)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @tbg)



<!-- Sent from Reviewable.io -->

Commit:
rpc,base: move HeartbeatInterval from const into Config

This enables dramatically shortening the runtime of the nodedialer tests from
over 3s to less than 50ms.

Release note: None

ajwerner commented:
Okay, added a second commit which shortens the test.

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
I hope the short heartbeat interval doesn't interfere in the tests in which you don't expect things to turn unhealthy.

> Reviewed 2 of 2 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35433#-:-L_Rd1orEZBYNgAeZLFR:bf5d7j6)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

ajwerner commented:
> I hope the short heartbeat interval doesn't interfere in the tests in which you don't expect things to turn unhealthy. 

I ran it in stress and stressrace for a bit and it seems happy. 

bors r+

ajwerner commented:
bors r+
